bioptim/optimization/solution.py

```diff
@@ -624,13 +626,22 @@ def __perform_integration(
 
                     else:
                         if keep_intermediate_points:
-                            integrated = np.array(nlp.dynamics[s](x0=x0, p=u, params=params / param_scaling)["xall"])
+                            if isinstance(nlp.ode_solver, OdeSolver.CVODES):
+                                raise NotImplementedError("keep_intermediate_points is not implemented with CVODES")
+                            else:
+                                integrated = np.array(nlp.dynamics[s](x0=x0, p=u, params=params / param_scaling)["xall"])
                             cols_in_out = [s * n_steps, (s + 1) * n_steps]
                         else:
-                            integrated = np.concatenate(
-                                (x0[:, np.newaxis], nlp.dynamics[s](x0=x0, p=u, params=params / param_scaling)["xf"]),
-                                axis=1,
-                            )
+                            if isinstance(nlp.ode_solver, OdeSolver.CVODES):
+                                integrated = np.concatenate(
+                                    (x0[:, np.newaxis], nlp.dynamics[s](x0=x0, p=u)["xf"]),
+                                    axis=1,
+                                )
+                            else:
+                                integrated = np.concatenate(
+                                    (x0[:, np.newaxis], nlp.dynamics[s](x0=x0, p=u, params=params / param_scaling)["xf"]),
+                                    axis=1,
+                                )
```

Comment:
As for the constraint, this if-statement should completely be removed and replaced by a proper interface to nlp.dynamics[s]